### ISSUE TITLE
(3.0.62) odd (but legal) class hierarchy causes trouble for Project Browser

### ISSUE THREAD

**Issue by dalehenrich** [ericwinger/Jade#442]
Here's the test method that produces the "odd (but legal) class hierarchy, with two `halts` one after initial structure is created and second after odd structure created:
```smalltalk
testXXY

  | projectName  packageName1 packageName2 projectDefinition classDefinition packageDefinition className1 className2 className3
    className4 projectSetDefinition class1 class2 class3 class4 oldClass1 oldClass2 oldClass3  |

  projectName := 'Issue470'.
  packageName1 := 'Issue470-Core'.
  packageName2 := 'Issue470-Extensions'.
  className1 := 'Issue470Class1'.
  className2 := 'Issue470Class2'.
  className3 := 'Issue470Class3'.
  className4 := 'Issue470Class4'.

  {projectName}
    do: [ :pn |
      (Rowan image loadedProjectNamed: pn ifAbsent: [  ])
        ifNotNil: [ :loadedProject | Rowan image _removeLoadedProject: loadedProject ] ].

"create project"
  projectDefinition := (RwProjectDefinition
    newForGitBasedProjectNamed: projectName)
    addPackageNamed: packageName1;
    addPackageNamed: packageName2;
    defaultSymbolDictName: self _symbolDictionaryName1;
    yourself.

  packageDefinition := projectDefinition packageNamed: packageName1.

  classDefinition := (RwClassDefinition
    newForClassNamed: className1
      super: 'Object'
      instvars: #(ivar1)
      classinstvars: #()
      classvars: #()
      category: packageName1
      comment: 'comment'
      pools: #()
      type: 'normal')
    yourself.
  packageDefinition
    addClassDefinition: classDefinition.

  classDefinition := (RwClassDefinition
    newForClassNamed: className2
      super: className1
      instvars: #('ivar2')
      classinstvars: #()
      classvars: #()
      category: packageName1
      comment: 'comment'
      pools: #()
      type: 'normal')
    yourself.
  packageDefinition
    addClassDefinition: classDefinition.

  classDefinition := (RwClassDefinition
    newForClassNamed: className3
      super: className2
      instvars: #('ivar4' 'ivar3')
      classinstvars: #()
      classvars: #()
      category: packageName1
      comment: 'comment'
      pools: #()
      type: 'normal')
   yourself.
  packageDefinition
    addClassDefinition: classDefinition.

"load"
  projectSetDefinition := RwProjectSetDefinition new.
  projectSetDefinition addDefinition: projectDefinition.
  Rowan projectTools load loadProjectSetDefinition: projectSetDefinition.

"validate"
  class1 := Rowan globalNamed: className1.
  self assert: class1 instVarNames = #(ivar1).

  class2 := Rowan globalNamed: className2.
  self assert: class2 instVarNames = #(ivar2).
  self assert: class2 superClass == class1.

  class3 := Rowan globalNamed: className3.
  self assert: class3 instVarNames = #(ivar4 ivar3).
  self assert: class3 superClass == class2.

self halt.
"remove class2 and add class4 -- edit projectDefinition structure in place"
  projectDefinition := (Rowan image loadedProjectNamed: projectName) asDefinition.
  packageDefinition := projectDefinition packageNamed: packageName1.

  packageDefinition removeClassNamed: className2.

  classDefinition := (RwClassDefinition
    newForClassNamed: className4
      super: className1
      instvars: #('ivar2')
      classinstvars: #()
      classvars: #()
      category: packageName1
      comment: 'comment'
      pools: #()
      type: 'normal')
    yourself.
  packageDefinition
    addClassDefinition: classDefinition.
"load"
  projectSetDefinition := RwProjectSetDefinition new.
  projectSetDefinition addDefinition: projectDefinition.
  Rowan projectTools load loadProjectSetDefinition: projectSetDefinition.

"validate"
  oldClass1 := class1.
  oldClass2 := class2.
  oldClass3 := class3.
 
  class1 := Rowan globalNamed: className1.
  self assert: class1 instVarNames = #(ivar1).
  self assert: oldClass1 == class1.

  class4 := Rowan globalNamed: className4.
  self assert: class4 instVarNames = #(ivar2).
  self assert: class4 superClass == class1.

  class3 := Rowan globalNamed: className3.
  self assert: class3 instVarNames = #(ivar4 ivar3).
  self assert: class3 superClass == oldClass2.
  self assert: oldClass3 == class3.

self halt.
```
The is screen shot after initial structure with three classes in a hierarchy are created:
![image](https://user-images.githubusercontent.com/623951/52735591-d7886080-2f7c-11e9-896e-20c0e4430bfc.png)
The odd structure is created by removing `Issue470Class2` and adding `Issue470Class4` .... without changing the superclass of `Issue470Class3`. This is the non-hierarchical view showing that the superclass of `Issue470Class3` is `Issue470Class2`:
![image](https://user-images.githubusercontent.com/623951/52735684-030b4b00-2f7d-11e9-9ab7-c6e1cfd311ea.png)
It's cool that the class `Issue470Class2` shows as the superclass of `Issue470Class`, but the fun starts when you select the hierarchy tab:
![image](https://user-images.githubusercontent.com/623951/52736035-bbd18a00-2f7d-11e9-929b-3da106b86be9.png)
I assume that this is similar to a case where you are looking at a subclass of a class that is not in your current symbol list (which is legal), so we should be able to handle this without errors ...

Definitely not a Feb bug, but in long run we should handle this case ... the Dictionary browser misbehaves as well, but not worth fixing, although integrated Dictionary view should hadle this case, too

**Comment by ericwinger** [comment]
The trouble starts when ``Rowan globalNamed: 'Issue470Class2'`` returns nil because, as you presumed, the class is not in ``GsCurrentSession currentSession symbolList``

The question is ... what should be displayed?

**Comment by dalehenrich** [comment]
If I'm not mistaken, I think that this is a case where you have a reference to a class that is not in your current symbol list ... and I think that the class still needs to be shown as a special case ... in tODE, I think I display this class as: `class-name{class oop}` ... we need to talk in person to resolve this correctly

**Comment by ericwinger** [comment]
From discussion with Dale, even though the non-resolvable class isn't in the symbol list, it is a fully formed class. Instead of resolving the class with ``Rowan globalNamed:`` we can get it by oop (if necessary, we may already have it). 

On the client we should give some indication, perhaps {oop} as was mentioned, that distinguishes it as a unique class.

**Comment by ericwinger** [comment]
Fixed in 3.0.69.  Hierarchy list looks as below with non-resolvable class in dark red and the oop beside it. 

![image](https://user-images.githubusercontent.com/36212516/57470517-dbefad00-723d-11e9-8293-4e7be76799aa.png)
